### PR TITLE
fix: Some months in the Persian Calendar show the 30th as the first day of the month

### LINKED ISSUE #5960: Some months in the Persian Calendar show the 30th as the first day of the month
### Provide a general summary of the issue here

image

As can be seen in the image above, the month shows the 30th as the first day of said month. The month before this month is also shown as "Esfand 1403 AP" and the following months (Farvardin, Ordibehesht, etc) of the new year (1404 AP) all have this issue until the year after that (1405 AP)

### 🤔 Expected Behavior?

The months should always start with the first I believe

### 😯 Current Behavior

See summary

### 💁 Possible Solution

_No response_

### 🔦 Context

_No response_

### 🖥️ Steps to Reproduce

Go to https://reactspectrum.blob.core.windows.net/reactspectrum/a0f0604e6dc461cc50dd923b494b7ad9b9e5bd0c/storybook/index.html?path=/story/date-and-time-calendar--default&providerSwitcher-express=false. Change the Calendar to Persian via the dropdown and navigate to the year 1404 AP. Note that the months have their start date as 31/30/29 depending on which month you are on.

### Version

3.34.1

### What browsers are you seeing the problem on?

Firefox, Chrome, Safari, Microsoft Edge

### If other, please specify.

_No response_

### What operating system are you using?

MacOS 14.1.1

### 🧢 Your Company/Team

RSP

### 🕷 Tracking Issue

_No response_

### PR DESCRIPTION
**Pull request by epr3**
Closes  #5960, #3757.

Tagging along this PR for context: #6035.

Hello, 

In collaboration with @sadeghbarati, we've found this library: https://github.com/jalaali/jalaali-js which outputs the correct months and month composition for Persian Calendar dates, thus I've ported the Julian Day conversion functions (and some utilities) plus the leap year calculations here and updated the `getYearsinEra` method.

I've also signed the CLA.

I'm looking forward to your feedback and any improvement suggestions.

Thanks!

## ✅ Pull Request Checklist:

- [x] Included link to corresponding [React Spectrum GitHub Issue](https://github.com/adobe/react-spectrum/issues).
- [ ] Added/updated unit tests and storybook for this change (for new code or code which already has tests).
- [ ] Filled out test instructions.
- [ ] Updated documentation (if it already exists for this component).
- [ ] Looked at the Accessibility Practices for this feature - [Aria Practices](https://www.w3.org/WAI/ARIA/apg/)

### CODE REVIEW

**Comment by sadeghbarati** [comment]
This PR also fixed inconsistently between "Jalali Datepickers"

In particular year like `1404/1/1` some of Jalali Datepickers first day of year start from `Friday` ✅ and in some others start from `Thursday` ❌  


<details><summary>Correct ✅ </summary>
<p>

Quasar Persian Datepicker
https://codepen.io/rstoenescu/pen/MWKpbNa

Screenshot 2024-03-19 at 1 06 41 PM


</p>
</details> 


<details><summary>Correct ✅ </summary>
<p>

https://www.time.ir/

Screenshot 2024-03-19 at 1 14 06 PM


</p>
</details> 



<details><summary>Incorrect ❌ </summary>
<p>

https://saeedrahimi.github.io/antd-jalali/

Screenshot 2024-03-19 at 1 16 06 PM


</p>
</details>

**Comment by LFDanLu** [comment]
Thanks for the fix! I'll see if I can take a look soon (probably will mainly be comparing against https://github.com/unicode-org/icu/blob/main/icu4j/main/core/src/main/java/com/ibm/icu/util/PersianCalendar.java#L422-L445 for the most part), but just want to let you know that the team might be a bit slow in reviewing due to some release prep we are handling, apologies in advance!

File: packages/@internationalized/date/tests/manipulation.test.js [hunk 1]
```diff
@@ -226,9 +226,9 @@ describe('CalendarDate manipulation', function () {
         expect(date.subtract({months: 1})).toEqual(new CalendarDate(new PersianCalendar(), 1, 1, 1));
       });
 
-      it('should constrain when reaching year 9377', function () {
-        let date = new CalendarDate(new PersianCalendar(), 9377, 12, 10);
-        expect(date.add({months: 1})).toEqual(new CalendarDate(new PersianCalendar(), 9377, 12, 31));
+      it('should constrain when reaching year 3178', function () {
+        let date = new CalendarDate(new PersianCalendar(), 3178, 12, 10);
+        expect(date.add({months: 1})).toEqual(new CalendarDate(new PersianCalendar(), 3178, 12, 31));
```
Comment (devongovett) [inline packages/@internationalized/date/tests/manipulation.test.js:231]: I'm not sure this is quite right. We generally want to match the output that Temporal, the upcoming JS standard, produces. If you try on https://tc39.es/proposal-temporal/docs/plaindate.html:

```js
let date = Temporal.PlainDate.from({year: 3178, month: 12, day: 10, calendar: 'persian'}).add({months: 1});
date.year // => 3179
date.month // => 1
date.day // => 10
```

That date converts to 3780-03-29 Gregorian, which formats using Intl.DateTimeFormat as a valid date in the Persian calendar.

```js
Intl.DateTimeFormat('en-US-u-ca-persian').format(new Date(3780, 3 - 1, 29))
// => '1/10/3159 AP'
```

I guess this was due to the change in `getYearsInEra`, which results in the date being constrained to this maximum. Temporal and Intl.DateTimeFormat don't seem to have this limit so I'm not sure we should either.

**Comment by devongovett** [comment]
Thanks so much for looking into this! I've opened a new PR in #6118 which simplifies this logic a bit, following how ICU is implemented. I've copied the tests over from various PRs/issues as well, and addressed my comment above. Would be great if you could take a look and see if you spot any remaining issues.

**Pull request closed without merging**